### ISSUE TITLE
Refactor of `PyLinter.stats`

### ISSUE THREAD

**Issue by DanielNoord** [PyCQA/pylint#5014]
### Current problem

The current problem is that the `stats` attribute of the main linter/checker classes support a wide array of different items. As of now the typing is:
https://github.com/PyCQA/pylint/blob/d4e61b9024dd6f073739fc6f7e6ecef2dceb9aaa/pylint/typing.py#L12-L14

Problems we are running into is stuff like this in `pylint/lint/pylinter.py`:
```python
def set_current_module(self, modname, filepath: Optional[str] = None):
        """set the name of the currently analyzed module and
        init statistics for it
        """
        if not modname and filepath is None:
            return
        self.reporter.on_set_current_module(modname, filepath)
        self.current_name = modname
        self.current_file = filepath or modname
        self.stats["by_module"][modname] = {}  # type: ignore # Refactor of PyLinter.stats necessary
        self.stats["by_module"][modname]["statement"] = 0  # type: ignore
        for msg_cat in MSG_TYPES.values():
            self.stats["by_module"][modname][msg_cat] = 0  # type: ignore
```
We can't specify the type of `by_module` which then causes issues with `mypy`.

I have tinkered around for a bit, but did not find a good solution/approach to this refactor. (I believe) it is impossible to type certain keys of a `Dict`/`TypedDict` and leave others unspecified. 
We would thus need a complete rework of the `stats` handling. This shouldn't be too big of an issue, but I would like some input on what maintainers/users think will be the best approach to this refactor and how they would ideally see `linter.stats` behave.

### Desired solution

/

**Comment by Pierre-Sassoulas** [comment]
I think we need to look at what is required and include that in a properly initialized and easily usable class. For example there's stat by module, class, methods... I think there might be some work with some design to do first.

**Comment by DanielNoord** [comment]
The most logical approach seems to be to use a `TypeDict` or `NamdeTuple`. However, we tend to loop over `nodes` or `modules` and add stats based on a concatenation of the node-variables and a prefix. This doesn't work with `TypedDict` as it requires string literals, while concatenation always returns `str`. See the following example:

```python
from typing import TypedDict

class Stats(TypedDict):
    badnames_ClassDef: int
    badnames_Module: int

stats = Stats(badnames_ClassDef=1, badnames_Module=1)

for node in ("ClassDef", "Module"):
    stats["badnames_" + node] = 1 # TypedDict key must be a string literal; expected one of ("badnames_ClassDef", "badnames_Module")
```

I have looked online but there doesn't seem to be a good way to resolve this. (Using `Final` is impossible since it is a loop).

They only thing I came up with so far is to make sure that `CheckerStats` always gets a `Dict` at the first level, so make it something like:
`Dict[str, Dict[str, Union[int, str, Counter[str], List, Dict[str, int]]]]`
Ideally we would also remove the `str`, `Counter` and `List` so it would become:
`Dict[str, Dict[str, Union[int, Dict[str, int]]]]`

However, we would then still have the problem of inferring the type of `stats["by_module"]`. Therefore, I'm open to better suggestions!

**Comment by Pierre-Sassoulas** [comment]
I'm for not using TypedDict at all. Right now we have a dict with possibly some key, like : ``nb_duplicated_line``, ``pct_duplicated_line``,  ``module``, ``class``, ``variable``. Or something dynamic like:
```python
        nice_stats[node_type] = {}
        if total != 0:
            try:
                undocumented_node: int = stats["undocumented_" + node_type]  # type: ignore
                documented = total - undocumented_node
                percent = (documented * 100.0) / total
                nice_stats[node_type]["percent_documented"] = f"{percent:.2f}"
            except KeyError:
                nice_stats[node_type]["percent_documented"] = "NC"
            try:
                badname_node: int = stats["badname_" + node_type]  # type: ignore
                percent = (badname_node * 100.0) / total
                nice_stats[node_type]["percent_badname"] = f"{percent:.2f}"
            except KeyError:
                nice_stats[node_type]["percent_badname"] = "NC"
```

You never know what's in this dict and the typing can only help so much, Beside there's a lot of polluting code in the checker for something that is specific to the stats. And it looks like the old stats are formatted into new stats when everything could be handled in some nice stats from the beginning. I'd do a class with helper function to remove that complexity from the checkers:

```python
class CheckerStats:

    """Small limited example for documentation of module"""

    def __init__():
        self.number_of_module; int = 0
        self.number_of_undocumented_module; int = 0

    def add_module(is_documented: bool=False):
         self.number_of_module += 1
         if not is_documented:
             self.number_of_undocumented_module += 1

    def percent_documented():
        return self.number_of_module - self.number_of_undocumented_module *100.0 / self.number_of_module

    def __str__():
        return # some usable representation for reporters maybe ?
```
